src/vdaf/mastic.rs

```diff
@@ -0,0 +1,727 @@
+        }
+    }
+
+    pub(crate) fn bits(&self) -> usize {
+        self.bits
+    }
+
+    pub(crate) fn vidpf(&self) -> &Vidpf<VidpfWeight<T::Field>, 16> {
+        &self.vidpf
+    }
+}
+
+/// Mastic aggregation parameter.
+///
+/// This includes the VIDPF tree level under evaluation and a set of prefixes to evaluate at that level.
+pub type MasticAggregationParam = Poplar1AggregationParam;
+
+/// Contains broadcast information shared between parties to support VIDPF correctness.
+pub type MasticPublicShare<V> = VidpfPublicShare<V>;
+
+impl<T, P, const SEED_SIZE: usize> ParameterizedDecode<Mastic<T, P, SEED_SIZE>>
+    for MasticPublicShare<VidpfWeight<T::Field>>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        decoding_parameter: &Mastic<T, P, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        MasticPublicShare::<VidpfWeight<T::Field>>::decode_with_param(
+            &(
+                decoding_parameter.bits(),
+                *decoding_parameter.vidpf().weight_parameter(),
+            ),
+            bytes,
+        )
+    }
+}
+
+/// Message sent by the [`Client`] to each Aggregator during the Sharding phase.
+#[derive(Clone, Debug)]
+pub struct MasticInputShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// VIDPF key share.
+    vidpf_key: VidpfKey,
+
+    /// The proof share.
+    proofs_share: SzkProofShare<F, SEED_SIZE>,
+}
+
+impl<'a, T, P, const SEED_SIZE: usize> ParameterizedDecode<(&'a Mastic<T, P, SEED_SIZE>, usize)>
+    for MasticInputShare<T::Field, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        (mastic, role): &(&'a Mastic<T, P, SEED_SIZE>, usize),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        let vidpf_key = VidpfKey::decode(bytes)?;
+        let proofs_share = SzkProofShare::<T::Field, SEED_SIZE>::decode_with_param(
+            &(*role == 0, mastic.proof_len(), mastic.is_randomized()),
+            bytes,
+        )?;
+        Ok(Self {
+            vidpf_key,
+            proofs_share,
+        })
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticInputShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.vidpf_key.encode(bytes)?;
+        self.proofs_share.encode(bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.vidpf_key.encoded_len()? + self.proofs_share.encoded_len()?)
+    }
+}
+
+#[cfg(test)]
+impl<F: FieldElement, const SEED_SIZE: usize> PartialEq for MasticInputShare<F, SEED_SIZE> {
+    fn eq(&self, other: &MasticInputShare<F, SEED_SIZE>) -> bool {
+        self.vidpf_key == other.vidpf_key && self.proofs_share == other.proofs_share
+    }
+}
+
+/// Struct containing a vector of VIDPF outputs: one for each prefix.
+#[derive(Clone, Debug)]
+pub struct MasticOutputShare<V: VidpfValue> {
+    result: Vec<V>,
+}
```

Comment:
Technically yes, but we still have to reimplement most of the things since VDAF only implements them for vectors of single field elements and we have nested vectors.